### ISSUE TITLE
with-emotion example with preact alias producing error in production app

### ISSUE THREAD

**Issue by vinaypuppal** [zeit/next.js#3315]
- [x ] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Context
I combined [with-emotion](https://github.com/zeit/next.js/tree/canary/examples/with-emotion) example and [using-preact](https://github.com/zeit/next.js/tree/canary/examples/using-preact) example. For production build I aliased `react` to `preact` as shown in [using-preact](https://github.com/zeit/next.js/tree/canary/examples/using-preact) example. After building the app(`next build`) and launching the app (`next start`) I am seeing below error in browser console.

<details>
 <summary>Stack Trace</summary>

```
_.default.hydrate is not a function
TypeError: _.default.hydrate is not a function
    at o (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:55939)
    at http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:57828
    at r (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:60761)
    at Generator._invoke (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:61806)
    at Generator.e.(anonymous function) [as next] (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:60940)
    at r (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:29712)
    at http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:29860
    at new Promise (<anonymous>)
    at new t (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:538)
    at http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:29654
(anonymous) @ app.js:2
r @ app.js:2
(anonymous) @ app.js:2
e.(anonymous function) @ app.js:2
r @ app.js:2
(anonymous) @ app.js:2
t @ app.js:2
(anonymous) @ app.js:2
(anonymous) @ app.js:2
(anonymous) @ app.js:2
r @ app.js:2
(anonymous) @ app.js:2
e.(anonymous function) @ app.js:2
r @ app.js:2
(anonymous) @ app.js:2
Promise rejected (async)
r @ app.js:2
(anonymous) @ app.js:2
t @ app.js:2
(anonymous) @ app.js:2
(anonymous) @ app.js:2
(anonymous) @ app.js:2
r @ app.js:2
(anonymous) @ app.js:2
e.(anonymous function) @ app.js:2
r @ app.js:2
(anonymous) @ app.js:2
Promise resolved (async)
r @ app.js:2
(anonymous) @ app.js:2
Promise resolved (async)
r @ app.js:2
(anonymous) @ app.js:2
t @ app.js:2
(anonymous) @ app.js:2
(anonymous) @ app.js:2
n @ app.js:1
(anonymous) @ app.js:2
n @ app.js:1
window.webpackJsonp @ app.js:1
(anonymous) @ app.js:3
app.js:2 Uncaught (in promise) TypeError: _.default.hydrate is not a function
    at o (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:55939)
    at http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:57048
    at r (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:60761)
    at Generator._invoke (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:61806)
    at Generator.e.(anonymous function) [as next] (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:60940)
    at r (http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:29712)
    at http://localhost:3000/_next/255b4ec75e53927fc721110ac93130bd/app.js:2:29808
    at <anonymous>
```
</details>

## Expected Behavior
Should not produce above error with `preact`. (It does not produce this error with `react` in production).

## Current Behavior
In production app its producing error mentioned in above stack trace.

## Steps to Reproduce (for bugs)
1. Download [with-emotion](https://github.com/zeit/next.js/tree/canary/examples/with-emotion) example (instructions in its Readme)
1. Update `next`, `react` and `react-dom`  to latest versions as versions used in this example are old.

    ```
     yarn add next react react-dom
    ```

1. Install `preact` and `preact-compat` 

    ```
    yarn add preact preact-compat
   ```

1. Create `next.config.js` file at root of the project and add below content 

     ```js
     module.exports = {
       webpack: function (config, { dev }) {
        // For the development version, we'll use React.
        // Because, it supports react hot loading and so on.
        if (dev) {
          return config
        }

        config.resolve.alias = {
          'react': 'preact-compat/dist/preact-compat',
          'react-dom': 'preact-compat/dist/preact-compat'
        }

         return config
      }
     }
      ```

1. Then build and start the app

     ```
     yarn build
    yarn start
    ```

1. Open app `https://localhost:3000` and check browsers console.

## Your Environment

| Tech    | Version |
|---------|---------|
| next    |    ^4.1.4     |
| node    |   v8.6.0      |
| OS      |     OSX 10.13.1    |
| browser |    chrome Version 64.0.3273.0 (Official Build) canary (64-bit)     |
| preact     |   ^8.2.6      |
| preact-compat | ^3.17.0 |
|react|^16.1.1|
|react-dom|^16.1.1|
|emotion| ^8.0.5 |

/cc @mitchellhamilton @tkh44

**Comment by alubbe** [comment]
I think it might be because @vinaypuppal did not add the module-alias to his server config.

The same error occurs when running the example repo in production, after updating to react 16 and preact 8 and disabling the module-alias:
https://github.com/zeit/next.js/tree/canary/examples/using-preact

Steps to reproduce:
```
curl https://codeload.github.com/zeit/next.js/tar.gz/master | tar -xz --strip=2 next.js-master/examples/using-preact
cd using-preact
# edit package.json to update react to 16 and preact to 8, as in the comment above
NODE_ENV=production npm run build
NODE_ENV=production npm run start
```

Nonetheless, with the module-alias, next refuses to run at all, not accepting preact:
```
Error: Next.js 4 requires React 16.
```

How can preact be used with next nowadays?

**Comment by alubbe** [comment]
Never mind, it has nothing to do module-alias, the reported error occurs either way. Help would be appreciated!

**Comment by alubbe** [comment]
Just ran the same test against using-inferno, with similar errors.

I might be going in circles, but maybe it is module-alias after all. I've only just realized that you need to call it *before* requiring `next` - how do you guys plan to do this when `import` becomes the standard way of doing this as opposed to `require`?

So my guess for this issue is that without the alias, the SSR is broken and therefore the rehydration fails on the client. Can preact (or inferno) be used with next 4?

**Comment by alubbe** [comment]
I guess this ticket might end up being a duplicate of https://github.com/zeit/next.js/issues/3264

Nonetheless, any plans for supporting `import`?

**Comment by vinaypuppal** [comment]
I just found this issue comment https://github.com/developit/preact-compat/issues/432#issuecomment-337361519 . So preact is not yet compatible with new React 16 API which `next@4.1.4` uses so I am getting above mentioned error.

And from readme
> Next.js 4 only supports React 16.
We had to drop React 15 support due to the way React 16 works and how we use it.

As this issue cannot be resolved on next.js side, so I am closing this issue.
